pkg/maps/nat/stats/stats.go

```diff
@@ -0,0 +1,325 @@
+			if m.natMap4 == nil && m.natMap6 == nil {
+				return nil
+			}
+
+			params.Jobs.Add(job.Timer("nat-stats", m.countNat, params.Config.NATMapStatInterval))
+			return params.Jobs.Start(hc)
+		},
+	})
+	return m, nil
+}
+
+func upsertStat[KT tupleKey](m *Stats, topk *topk[KT], family nat.IPFamily) error {
+	tx := m.db.WriteTxn(m.table)
+	defer tx.Abort()
+
+	var errs error
+	iter, _ := m.table.All(tx)
+	for entry, _, ok := iter.Next(); ok; entry, _, ok = iter.Next() {
+		if entry.Type == family.String() {
+			_, _, err := m.table.Delete(tx, entry)
+			errors.Join(errs, err)
+		}
+	}
+
+	topk.popForEach(func(key KT, count, ith int) {
+		if ith == topk.size-1 {
+			m.metrics.updateLocalPorts(family, count, m.maxPorts)
+		}
+		extip, eip, proto, rport := key.tuple()
+		_, _, err := m.table.Insert(tx, NatMapStats{
+			Type:       family.String(),
+			EgressIP:   eip,
+			EndpointIP: extip,
+			RemotePort: rport,
+			Proto:      proto,
+			Count:      count,
+			Nth:        ith,
+		})
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+	})
+	if errs != nil {
+		return fmt.Errorf("failures occured updating stats table, transaction will not be committed: %w", errs)
+	}
+	tx.Commit()
+	return nil
+}
+
+func (m *Stats) run(ctx context.Context, hr cell.Health) {
+	controller.NewManager().UpdateController("nat-stats", controller.ControllerParams{
+		RunInterval: m.config.NATMapStatInterval,
+		Context:     ctx,
+		Health:      hr,
+		Group:       controller.NewGroup("nat-stats"),
+		DoFunc:      m.countNat,
+	})
+}
+
+func swapBytes(n uint16) uint16 {
+	return (n<<8)&0xff00 | (n>>8)&0x00ff
+}
+
+func (m *Stats) countNat(ctx context.Context) error {
+	var errs error
+	if m.natMap4 != nil {
+		tupleToPortCount := make(map[tuple.TupleKey4]uint16, 128)
+		_, err := m.natMap4.ApplyBatch4(func(keys []tuple.TupleKey4, vals []nat.NatEntry4, size int) {
+			for i := 0; i < size; i++ {
+				key := keys[i]
+				key.SourcePort = swapBytes(key.SourcePort)
+				if key.Flags == tuple.TUPLE_F_IN &&
+					(key.NextHeader == u8proto.TCP || key.NextHeader == u8proto.ICMP ||
+						key.NextHeader == u8proto.UDP) {
+					key.DestPort = 0
+					ports := tupleToPortCount[key]
+					ports++
+					tupleToPortCount[key] = ports
+				}
+			}
+		})
+		if err != nil {
+			m.logger.WithError(err).
+				Error("failed to count ipv4 nat map entries, " +
+					"this may result in out of date nat-stats data and nat_endpoint_ metrics")
+			errs = errors.Join(errs, err)
+		} else {
+			topk := newTopK[key4](m.config.NatMapStatKStoredEntries)
+			for tupleKey, bucket := range tupleToPortCount {
+				topk.Push(key4(tupleKey), int(bucket))
+			}
+			errors.Join(errs, upsertStat(m, topk, nat.IPv4))
+		}
+	}
+	if m.natMap6 != nil {
+		tupleToPortCount := make(map[tuple.TupleKey6]uint16, 128)
+		_, err := m.natMap6.ApplyBatch6(func(keys []tuple.TupleKey6, vals []nat.NatEntry6, size int) {
+			for i := 0; i < size; i++ {
+				key := keys[i]
+				key.SourcePort = swapBytes(key.SourcePort)
```

Comment:
Not knowing the details, but this seems wrong. Do we know for a fact the source and target endianess? Shouldn't this use `binary.BigEndian.PutUint16` and friends?

Reply:
These are inserting by the datapath directly from packet ctx, thus they are in network order.  However I should be using the `tuple.MapKey.(ToHost)` func to make this clear and for code reuse 👍